`nested-min-max` (W3301) is still giving a wrong suggestion in some cases

### Bug description

With the following code
```python
""" An example where pylint gives a wrong suggestion """
max(3, max([5] + [i for i in range(6,10) if i % 2 == 0]))
```
pylint detects a nested call of `max` and suggests to just remove the inner `max`.  But this leads to a type error:
```
TypeError: '>' not supported between instances of 'list' and 'int'
```

### Configuration

_No response_

### Command used

```shell
pylint a.py
```


### Pylint output

```shell
************* Module a
a.py:2:0: W3301: Do not use nested call of 'max'; it's possible to do 'max(3, [5] + [i for i in range(6, 10) if i % 2 == 0])' instead (nested-min-max)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```


### Expected behavior

At the very least not give a wrong suggestion, but ideally give a right one, like for other instances of this problem recently addressed in https://github.com/pylint-dev/pylint/issues/8168 and https://github.com/pylint-dev/pylint/pull/8234.

For instance, with
```python
max(3, max([5] + [4]))
```
we now get the working suggestion to replace it with
```python
max(3, *[5] + [4])
```
so we could suggest in this case too to add a `*`:
```python
max(3, *[5] + [i for i in range(6, 10) if i % 2 == 0])
```
(In this case, one could of course even be smarter, but this may be too specific.)

### Pylint version

```shell
pylint 2.17.0
astroid 2.15.0
Python 3.10.10 (main, Mar  5 2023, 22:26:53) [GCC 12.2.1 20230201]
```


### OS / Environment

_No response_

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/8524.bugfix
+++ b/doc/whatsnew/fragments/8524.bugfix
@@ -1,0 +1,3 @@
+Fixes suggestion for ``nested-min-max`` for expressions with additive operators, list and dict comprehensions.
+
+Closes #8524

--- a/pylint/checkers/nested_min_max.py
+++ b/pylint/checkers/nested_min_max.py
@@ -124,7 +124,7 @@ class NestedMinMaxChecker(BaseChecker):
         )
 
     def _is_splattable_expression(self, arg: nodes.NodeNG) -> bool:
-        """Returns true if expression under minmax could be converted to splat
+        """Returns true if expression under min/max could be converted to splat
         expression.
         """
         # Support sequence addition (operator __add__)

--- a/pylint/checkers/nested_min_max.py
+++ b/pylint/checkers/nested_min_max.py
@@ -14,6 +14,7 @@ from astroid.const import Context
 
 from pylint.checkers import BaseChecker
 from pylint.checkers.utils import only_required_for_messages, safe_infer
+from pylint.constants import PY39_PLUS
 from pylint.interfaces import INFERENCE
 
 if TYPE_CHECKING:
@@ -93,13 +94,10 @@ class NestedMinMaxChecker(BaseChecker):
 
         for idx, arg in enumerate(fixed_node.args):
             if not isinstance(arg, nodes.Const):
-                inferred = safe_infer(arg)
-                if isinstance(
-                    inferred, (nodes.List, nodes.Tuple, nodes.Set, *DICT_TYPES)
-                ):
+                if self._is_splattable_expression(arg):
                     splat_node = nodes.Starred(
                         ctx=Context.Load,
-                        lineno=inferred.lineno,
+                        lineno=arg.lineno,
                         col_offset=0,
                         parent=nodes.NodeNG(
                             lineno=None,
@@ -124,6 +122,39 @@ class NestedMinMaxChecker(BaseChecker):
             args=(node.func.name, fixed_node.as_string()),
             confidence=INFERENCE,
         )
+
+    def _is_splattable_expression(self, arg: nodes.NodeNG) -> bool:
+        """Returns true if expression under minmax could be converted to splat
+        expression.
+        """
+        # Support sequence addition (operator __add__)
+        if isinstance(arg, nodes.BinOp) and arg.op == "+":
+            return self._is_splattable_expression(
+                arg.left
+            ) and self._is_splattable_expression(arg.right)
+        # Support dict merge (operator __or__ in Python 3.9)
+        if isinstance(arg, nodes.BinOp) and arg.op == "|" and PY39_PLUS:
+            return self._is_splattable_expression(
+                arg.left
+            ) and self._is_splattable_expression(arg.right)
+
+        inferred = safe_infer(arg)
+        if inferred and inferred.pytype() in {"builtins.list", "builtins.tuple"}:
+            return True
+        if isinstance(
+            inferred or arg,
+            (
+                nodes.List,
+                nodes.Tuple,
+                nodes.Set,
+                nodes.ListComp,
+                nodes.DictComp,
+                *DICT_TYPES,
+            ),
+        ):
+            return True
+
+        return False
 
 
 def register(linter: PyLinter) -> None:

--- a/tests/functional/n/nested_min_max.py
+++ b/tests/functional/n/nested_min_max.py
@@ -54,6 +54,3 @@ max(3, *[5] + list(range(4)))
 
 max(3, max(list(range(4))))  # [nested-min-max]
 max(3, *list(range(4)))
-
-max(3, max({1: 2} | {i: i for i in range(4) if i}))  # [nested-min-max]
-max(3, *{1: 2} | {i: i for i in range(4) if i})

--- a/tests/functional/n/nested_min_max.py
+++ b/tests/functional/n/nested_min_max.py
@@ -42,3 +42,18 @@ max(3, *nums.values())
 
 lst2 = [3, 7, 10]
 max(3, max(nums), max(lst2))  # [nested-min-max]
+
+max(3, max([5] + [6, 7]))  # [nested-min-max]
+max(3, *[5] + [6, 7])
+
+max(3, max([5] + [i for i in range(4) if i]))  # [nested-min-max]
+max(3, *[5] + [i for i in range(4) if i])
+
+max(3, max([5] + list(range(4))))  # [nested-min-max]
+max(3, *[5] + list(range(4)))
+
+max(3, max(list(range(4))))  # [nested-min-max]
+max(3, *list(range(4)))
+
+max(3, max({1: 2} | {i: i for i in range(4) if i}))  # [nested-min-max]
+max(3, *{1: 2} | {i: i for i in range(4) if i})

--- a/tests/functional/n/nested_min_max.txt
+++ b/tests/functional/n/nested_min_max.txt
@@ -16,4 +16,3 @@ nested-min-max:46:0:46:25::Do not use nested call of 'max'; it's possible to do 
 nested-min-max:49:0:49:45::Do not use nested call of 'max'; it's possible to do 'max(3, *[5] + [i for i in range(4) if i])' instead:INFERENCE
 nested-min-max:52:0:52:33::Do not use nested call of 'max'; it's possible to do 'max(3, *[5] + list(range(4)))' instead:INFERENCE
 nested-min-max:55:0:55:27::Do not use nested call of 'max'; it's possible to do 'max(3, *list(range(4)))' instead:INFERENCE
-nested-min-max:58:0:58:51::"Do not use nested call of 'max'; it's possible to do 'max(3, *{1: 2} | {i: i for i in range(4) if i})' instead":INFERENCE

--- a/tests/functional/n/nested_min_max.txt
+++ b/tests/functional/n/nested_min_max.txt
@@ -12,3 +12,8 @@ nested-min-max:33:0:33:17::Do not use nested call of 'max'; it's possible to do 
 nested-min-max:37:0:37:17::Do not use nested call of 'max'; it's possible to do 'max(3, *nums)' instead:INFERENCE
 nested-min-max:40:0:40:26::Do not use nested call of 'max'; it's possible to do 'max(3, *nums.values())' instead:INFERENCE
 nested-min-max:44:0:44:28::Do not use nested call of 'max'; it's possible to do 'max(3, *nums, *lst2)' instead:INFERENCE
+nested-min-max:46:0:46:25::Do not use nested call of 'max'; it's possible to do 'max(3, *[5] + [6, 7])' instead:INFERENCE
+nested-min-max:49:0:49:45::Do not use nested call of 'max'; it's possible to do 'max(3, *[5] + [i for i in range(4) if i])' instead:INFERENCE
+nested-min-max:52:0:52:33::Do not use nested call of 'max'; it's possible to do 'max(3, *[5] + list(range(4)))' instead:INFERENCE
+nested-min-max:55:0:55:27::Do not use nested call of 'max'; it's possible to do 'max(3, *list(range(4)))' instead:INFERENCE
+nested-min-max:58:0:58:51::"Do not use nested call of 'max'; it's possible to do 'max(3, *{1: 2} | {i: i for i in range(4) if i})' instead":INFERENCE

--- a/tests/functional/n/nested_min_max_py39.py
+++ b/tests/functional/n/nested_min_max_py39.py
@@ -1,0 +1,6 @@
+"""Test detection of redundant nested calls to min/max functions"""
+
+# pylint: disable=redefined-builtin,unnecessary-lambda-assignment
+
+max(3, max({1: 2} | {i: i for i in range(4) if i}))  # [nested-min-max]
+max(3, *{1: 2} | {i: i for i in range(4) if i})

--- a/tests/functional/n/nested_min_max_py39.rc
+++ b/tests/functional/n/nested_min_max_py39.rc
@@ -1,0 +1,2 @@
+[testoptions]
+min_pyver=3.9

--- a/tests/functional/n/nested_min_max_py39.txt
+++ b/tests/functional/n/nested_min_max_py39.txt
@@ -1,0 +1,1 @@
+nested-min-max:5:0:5:51::"Do not use nested call of 'max'; it's possible to do 'max(3, *{1: 2} | {i: i for i in range(4) if i})' instead":INFERENCE

